README.md

```diff
@@ -65,6 +65,20 @@ KeycloakOauth.configure do |config|
 end
 ```
 
+**User mapping**
+The host app is responsible for mapping a Keycloak session with a Rails user session. This can be achieved
+by implementing the `map_authenticatable` method in the module configured above (e.g. `KeycloakOauthCallbacks` in our example).
+You can get the user information using the `KeycloakOauth::UserInfoRetrievalService` to which you pass in the access token.
+See here an example of retrieving the user information and saving the email address in the Rails session:
+
+```ruby
+def map_authenticatable(_request)
+  service = KeycloakOauth::UserInfoRetrievalService.new(access_token: session[:access_token])
+  service.retrieve
```

Comment:
I find it a bit strange that the `retreive` method has to be called separately to access the user information. Is that a ruby-esque thing?
I'd rather return the user information directly from the `retrieve` method (and call it `get_user_information`). What do you think?

Reply:
Do you mean having a module instead of a class, with a class method `get_user_information`?

I was thinking to handle the user information within a class, because I'm envisioning adding some accessors for the user information in the future e.g. `groups` (although I think this might go against the idea behind service objects, where you really just want to perform something and that's it, see an example here: https://www.toptal.com/ruby-on-rails/rails-service-objects-tutorial). 

We could perform the request in the initializer, but I'm not sure it's a good idea to make an external request when initialising the object (I feel like it hides this additional logic from the component calling it, whereas having to call `retrieve` makes it explicit that the call will have a side-effect).

Now I'm also thinking that user information could be treated as an entity/model on which we can call things like `groups`, `email`, etc. But I think there is still a need for a service, since we first need to perform an external request before we can initialise a user information model.

Maybe we could have something like this:
- from host app, call `KeycloakOauth.connection.get_user_information(access_token, options)` where options could have some filtering/scoping if needed).
- then, we'd implement the following:
```ruby
# in KeycloakOauth::Connection

def get_user_information(access_token, options)
   service = UserInformationRetrieval.new(access_token, options)
   service.retrieve
   UserInformation.new(service.user_info)
end
```

```ruby
class UserInformation
  def initialize(user_info)
    ...
  end

  def groups
    user_info['groups']
  end

  def email
  end

  ....
end
```

Then, in the host app, you'd only call `KeycloakOauth.connection.get_user_information(...)`

What do you think about this approach?

Reply:
Thanks for your detailed thoughts. I like the proposal, let's go with this.